pkg/odo/cli/registry/registry.go

```diff
@@ -0,0 +1,241 @@
+		Short:   "List all components from the Devfile registry",
+		Long:    "List all components from the Devfile registry",
+		Example: fmt.Sprintf(Example, fullName),
+		Run: func(cmd *cobra.Command, args []string) {
+			genericclioptions.GenericRun(o, cmd, args)
+		},
+	}
+
+	clientset.Add(listCmd, clientset.REGISTRY)
+
+	// Flags
+	listCmd.Flags().StringVar(&o.filterFlag, "filter", "", "Filter based on the name of the component")
+	listCmd.Flags().StringVar(&o.devfileFlag, "devfile", "", "Only the specific Devfile component")
+	listCmd.Flags().StringVar(&o.registryFlag, "devfile-registry", "", "Only show components from the specific Devfile registry")
+	listCmd.Flags().BoolVar(&o.detailsFlag, "details", false, "Show details of each component")
+
+	// Add a defined annotation in order to appear in the help menu
+	listCmd.Annotations["command"] = "main"
+	listCmd.SetUsageTemplate(odoutil.CmdUsageTemplate)
+
+	return listCmd
+}
+
+func (o *ListOptions) printDevfileList(DevfileList []registry.DevfileStack) {
+
+	// Create the table and use our own style
+	t := table.NewWriter()
+
+	t.SetStyle(table.Style{
+		Box: table.BoxStyle{
+			PaddingLeft:  " ",
+			PaddingRight: " ",
+		},
+		Color: table.ColorOptions{
+			Header: text.Colors{text.FgHiGreen, text.Underline},
+		},
+		Format: table.FormatOptions{
+			Footer: text.FormatUpper,
+			Header: text.FormatUpper,
+			Row:    text.FormatDefault,
+		},
+		Options: table.Options{
+			DrawBorder:      false,
+			SeparateColumns: false,
+			SeparateFooter:  false,
+			SeparateHeader:  false,
+			SeparateRows:    false,
+		},
+	})
+	t.SetOutputMirror(log.GetStdout())
+
+	t.AppendHeader(table.Row{"NAME", "REGISTRY", "DESCRIPTION"})
+
+	devfiles := []registry.DevfileStack{}
+
+	// Filter through all the devfile components per the filters / parameters passed in.
+	for _, devfileComponent := range DevfileList {
+
+		// If the user has specified a filter with variable o.filterFlag, then only show the components
+		// containing that specific string.
+		if o.filterFlag != "" {
+			if !strings.Contains(devfileComponent.Name, o.filterFlag) && !strings.Contains(devfileComponent.Description, o.filterFlag) {
+				continue
+			}
+		}
+
+		// If the user passed in --devfile <NAME> only show that specific component matching that name
+		if o.devfileFlag != "" {
+			if devfileComponent.Name != o.devfileFlag {
+				continue
+			}
+		}
+		devfiles = append(devfiles, devfileComponent)
+	}
+
+	for _, devfileComponent := range devfiles {
+		// Mark the name as yellow in the index so it's easier to see.
+		name := text.Colors{text.FgHiYellow}.Sprint(devfileComponent.Name)
+
+		if o.detailsFlag {
+
+			// If details are being shown, we must download the Devfile from the registry
+			// parse it, in order to retrieve command information such as if debug, dev, deploy and test is enabled.
+			devfileData, err := o.clientset.RegistryClient.RetrieveDevfileDataFromRegistry(devfileComponent.Registry.Name, devfileComponent.Name)
```

Comment:
We will also need this data for the JSON output, please move this code to the business layer

Reply:
Have to keep it in this layer.. because this is only triggered with `--details`. If we had this in the business layer, we would http get every single devfile which is what we don't want.

Or I could add it as a boolean of true/false for the function if we require details or not? What do you think?

Reply:
Yes, you could pass an option to the business layer function to indicate if you want or not the details